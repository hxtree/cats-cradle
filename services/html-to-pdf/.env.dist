AWS_ACCOUNT_ID=123456789012
AWS_REGION=us-east-1
STAGE_NAME=default
BASE_URL=

APP_NAME=html-to-pdf
ENVIROMENT=local

AWS_BUCKET=298217020330-default-html-to-pdf-bucket
AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=