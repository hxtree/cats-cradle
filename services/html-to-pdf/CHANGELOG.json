{
  "name": "@cats-cradle/html-to-pdf",
  "entries": [
    {
      "version": "1.0.13",
      "tag": "@cats-cradle/html-to-pdf_v1.0.13",
      "date": "Sun, 05 Nov 2023 05:26:39 GMT",
      "comments": {
        "patch": [
          {
            "comment": "add badge to readme"
          }
        ]
      }
    },
    {
      "version": "1.0.12",
      "tag": "@cats-cradle/html-to-pdf_v1.0.12",
      "date": "Sat, 04 Nov 2023 19:40:56 GMT",
      "comments": {
        "patch": [
          {
            "comment": "add stack snapshot tests"
          }
        ],
        "dependency": [
          {
            "comment": "Updating dependency \"@cats-cradle/constructs\" to `0.0.5`"
          },
          {
            "comment": "Updating dependency \"@cats-cradle/base-nodejs\" to `1.0.7`"
          },
          {
            "comment": "Updating dependency \"@cats-cradle/faker-factory\" to `1.2.4`"
          }
        ]
      }
    },
    {
      "version": "1.0.11",
      "tag": "@cats-cradle/html-to-pdf_v1.0.11",
      "date": "Fri, 03 Nov 2023 04:39:27 GMT",
      "comments": {
        "dependency": [
          {
            "comment": "Updating dependency \"@cats-cradle/base-nodejs\" to `1.0.5`"
          },
          {
            "comment": "Updating dependency \"@cats-cradle/faker-factory\" to `1.2.3`"
          }
        ]
      }
    },
    {
      "version": "1.0.10",
      "tag": "@cats-cradle/html-to-pdf_v1.0.10",
      "date": "Thu, 02 Nov 2023 22:51:17 GMT",
      "comments": {
        "dependency": [
          {
            "comment": "Updating dependency \"@cats-cradle/faker-factory\" to `1.2.2`"
          }
        ]
      }
    },
    {
      "version": "1.0.9",
      "tag": "@cats-cradle/html-to-pdf_v1.0.9",
      "date": "Thu, 02 Nov 2023 12:35:11 GMT",
      "comments": {
        "dependency": [
          {
            "comment": "Updating dependency \"@cats-cradle/faker-factory\" to `1.2.1`"
          }
        ]
      }
    },
    {
      "version": "1.0.8",
      "tag": "@cats-cradle/html-to-pdf_v1.0.8",
      "date": "Wed, 01 Nov 2023 04:48:36 GMT",
      "comments": {
        "dependency": [
          {
            "comment": "Updating dependency \"@cats-cradle/faker-factory\" to `1.2.0`"
          }
        ]
      }
    },
    {
      "version": "1.0.7",
      "tag": "@cats-cradle/html-to-pdf_v1.0.7",
      "date": "Sat, 28 Oct 2023 21:42:52 GMT",
      "comments": {
        "dependency": [
          {
            "comment": "Updating dependency \"@cats-cradle/faker-factory\" to `1.1.0`"
          }
        ]
      }
    },
    {
      "version": "1.0.6",
      "tag": "@cats-cradle/html-to-pdf_v1.0.6",
      "date": "Sat, 28 Oct 2023 04:52:47 GMT",
      "comments": {
        "patch": [
          {
            "comment": "change ci"
          }
        ]
      }
    },
    {
      "version": "1.0.5",
      "tag": "@cats-cradle/html-to-pdf_v1.0.5",
      "date": "Mon, 23 Oct 2023 12:23:42 GMT",
      "comments": {
        "patch": [
          {
            "comment": "Add repo info to package.json"
          }
        ]
      }
    }
  ]
}
