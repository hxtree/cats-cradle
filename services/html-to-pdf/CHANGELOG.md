# Change Log - @cats-cradle/html-to-pdf

This log was last generated on Thu, 28 Dec 2023 16:54:41 GMT and should not be manually modified.

## 1.0.18
Thu, 28 Dec 2023 16:54:41 GMT

### Patches

- upgrade typescript 5.3.3

## 1.0.17
Tue, 12 Dec 2023 03:15:41 GMT

### Patches

- add diagram

## 1.0.16
Sun, 03 Dec 2023 08:29:38 GMT

_Version update only_

## 1.0.15
Mon, 13 Nov 2023 02:11:53 GMT

### Patches

- update packages

## 1.0.14
Sun, 05 Nov 2023 18:41:09 GMT

_Version update only_

## 1.0.13
Sun, 05 Nov 2023 05:26:39 GMT

### Patches

- add badge to readme

## 1.0.12
Sat, 04 Nov 2023 19:40:56 GMT

### Patches

- add stack snapshot tests

## 1.0.11
Fri, 03 Nov 2023 04:39:27 GMT

_Version update only_

## 1.0.10
Thu, 02 Nov 2023 22:51:17 GMT

_Version update only_

## 1.0.9
Thu, 02 Nov 2023 12:35:11 GMT

_Version update only_

## 1.0.8
Wed, 01 Nov 2023 04:48:36 GMT

_Version update only_

## 1.0.7
Sat, 28 Oct 2023 21:42:52 GMT

_Version update only_

## 1.0.6
Sat, 28 Oct 2023 04:52:47 GMT

### Patches

- change ci

## 1.0.5
Mon, 23 Oct 2023 12:23:42 GMT

### Patches

- Add repo info to package.json

