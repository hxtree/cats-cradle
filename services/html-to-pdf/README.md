# @org-apis/html-to-pdf

HTMLtoPDF is a service for generating PDFs from HTML. It support generating a
PDF from either a request containing a HTML body or a URL.

## How it Works

PDF are generated using a headless version of Chromium. This form of PDF
rendering supports text recognition, images, hyperlinks, print media queries,
table breaks, and other features all this with relatively little code
maintenance.

It can be finicky to get working within a Lambda and API Gateway. Lambda doesn't
include fonts fonts. Chromium should be installed in a Lambda layer. API gateway
can cause blank PDF if it doesn't properly handle binary responses. Serverless
Express also has to be configured to support the binary mime type.

## References

- [Generating PDF AWS Lambda](https://wavelop.com/en/story/generate-pdf-with-aws-lambda/)
- [Puppeteer Layers](https://github.com/RafalWilinski/serverless-puppeteer-layers/tree/master/layer)
- [Binary Lambda Responses](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-payload-encodings-configure-with-console.html)
- [Serverless Express Binary](https://github.com/vendia/serverless-express/blob/master/examples/basic-starter/lambda.js)
