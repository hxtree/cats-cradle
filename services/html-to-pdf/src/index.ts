// used in lambda
import { VersioningType } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ExpressAdapter } from '@nestjs/platform-express';
import serverlessExpress from '@vendia/serverless-express';
import { Context, Handler } from 'aws-lambda';
import express from 'express';

import { AppModule } from './app.module';

let cachedServer: Handler;

async function bootstrap() {
  if (!cachedServer) {
    const expressApp = express();
    const nestApp = await NestFactory.create(
      AppModule,
      new ExpressAdapter(expressApp),
    );

    const contentTypes = [
      'application/pdf',
      // 'application/octet-stream',
      // 'font/eot',
      // 'font/opentype',
      // 'font/otf',
      // 'image/jpeg',
      // 'image/png',
      // 'image/svg+xml',
    ];

    nestApp.enableVersioning({
      type: VersioningType.HEADER,
      header: 'Accept-Version',
      defaultVersion: '1',
    });

    nestApp.enableCors();

    await nestApp.init();

    cachedServer = serverlessExpress({
      app: expressApp,
      binarySettings: { contentTypes },
    });
  }

  return cachedServer;
}

export const handler = async (event: any, context: Context, callback: any) => {
  const server = await bootstrap();
  return server(event, context, callback);
};
