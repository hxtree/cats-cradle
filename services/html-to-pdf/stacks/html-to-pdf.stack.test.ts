import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { HtmlToPdfStack } from './html-to-pdf.stack';
import { downloadChromiumZip } from './download-lambda-layer';

describe('HtmlToPdfStack', () => {
  beforeAll(async () => {
    await downloadChromiumZip();
  });

  it('should match snapshot test', () => {
    const app = new cdk.App();
    const stack = new HtmlToPdfStack(app, 'MyTestStack');
    const template = Template.fromStack(stack);

    // TODO consider reworking to use DNS instead of incrementing and deploying each time
    const customSnapshotSerializer = {
      test: (val: any) => typeof val === 'string',
      print: (val: any) => val
        .replace(/deployment\d{8}T\d{9}Z[^:]+/, 'deployment_UNIQUE_HASH')
        .replace(/[a-fA-F0-9]+\.zip/, 'UNIQUE_HASH.zip'),
    };
    expect.addSnapshotSerializer(customSnapshotSerializer);

    expect(template.toJSON()).toMatchSnapshot();
  });
});
