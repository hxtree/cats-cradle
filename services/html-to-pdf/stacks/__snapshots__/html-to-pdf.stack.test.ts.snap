// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlToPdfStack should match snapshot test 1`] = `
{
  Outputs: {
    exampletestendpoint: {
      Value: {
        Fn::Join: [
          ,
          [
            https://,
            {
              Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4,
            },
            /pdf?url=https://google.com,
          ],
        ],
      },
    },
    healthcheckendpoint: {
      Value: {
        Fn::Join: [
          ,
          [
            https://,
            {
              Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4,
            },
            /health,
          ],
        ],
      },
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiEndpoint26E247D6: {
      Value: {
        Fn::Join: [
          ,
          [
            https://,
            {
              Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
            },
            .execute-api.,
            {
              Ref: AWS::Region,
            },
            .,
            {
              Ref: AWS::URLSuffix,
            },
            /,
            {
              Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentStageprod256692E7,
            },
            /,
          ],
        ],
      },
    },
  },
  Parameters: {
    BootstrapVersion: {
      Default: /cdk-bootstrap/hnb659fds/version,
      Description: Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip],
      Type: AWS::SSM::Parameter::Value<String>,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointhostedzoneidParameter400E20C0: {
      Default: my-domains-hosted-zone-id,
      Type: AWS::SSM::Parameter::Value<String>,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointorgformationcertswildcardcert1arnParameter1AF8E75E: {
      Default: orgformation-certs-wildcard-cert1-arn,
      Type: AWS::SSM::Parameter::Value<String>,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointssmdomainnameParameterD95B813B: {
      Default: DOMAIN_NAME,
      Type: AWS::SSM::Parameter::Value<String>,
    },
    lambdalayernestjslatestversionssmParameter: {
      Default: lambda-layer-nestjs-latest-version,
      Type: AWS::SSM::Parameter::Value<String>,
    },
    mongodatabasepasswordssmParameter: {
      Default: MONGO_DATABASE_PASSWORD,
      Type: AWS::SSM::Parameter::Value<String>,
    },
    mongodatabaseurissmParameter: {
      Default: MONGO_DATABASE_URI,
      Type: AWS::SSM::Parameter::Value<String>,
    },
    mongodatabaseuserssmParameter: {
      Default: MONGO_DATABASE_USER,
      Type: AWS::SSM::Parameter::Value<String>,
    },
  },
  Resources: {
    ChromiumLayer505381F6: {
      Properties: {
        CompatibleRuntimes: [
          nodejs18.x,
        ],
        Content: {
          S3Bucket: {
            Fn::Sub: cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region},
          },
          S3Key: UNIQUE_HASH.zip,
        },
        Description: Chromium layer for Lambda,
        LayerName: ChromiumLayer,
      },
      Type: AWS::Lambda::LayerVersion,
    },
    LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A: {
      DependsOn: [
        LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB,
        LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB,
      ],
      Properties: {
        Code: {
          S3Bucket: {
            Fn::Sub: cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region},
          },
          S3Key: UNIQUE_HASH.zip,
        },
        Handler: index.handler,
        Role: {
          Fn::GetAtt: [
            LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB,
            Arn,
          ],
        },
        Runtime: nodejs18.x,
        Timeout: 900,
      },
      Type: AWS::Lambda::Function,
    },
    LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB: {
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Action: sts:AssumeRole,
              Effect: Allow,
              Principal: {
                Service: lambda.amazonaws.com,
              },
            },
          ],
          Version: 2012-10-17,
        },
        ManagedPolicyArns: [
          {
            Fn::Join: [
              ,
              [
                arn:,
                {
                  Ref: AWS::Partition,
                },
                :iam::aws:policy/service-role/AWSLambdaBasicExecutionRole,
              ],
            ],
          },
        ],
      },
      Type: AWS::IAM::Role,
    },
    LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB: {
      Properties: {
        PolicyDocument: {
          Statement: [
            {
              Action: [
                logs:PutRetentionPolicy,
                logs:DeleteRetentionPolicy,
              ],
              Effect: Allow,
              Resource: *,
            },
          ],
          Version: 2012-10-17,
        },
        PolicyName: LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB,
        Roles: [
          {
            Ref: LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB,
          },
        ],
      },
      Type: AWS::IAM::Policy,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointapigatewayaliasrecord366F7DAC: {
      Properties: {
        AliasTarget: {
          DNSName: {
            Fn::GetAtt: [
              htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4,
              RegionalDomainName,
            ],
          },
          HostedZoneId: {
            Fn::GetAtt: [
              htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4,
              RegionalHostedZoneId,
            ],
          },
        },
        HostedZoneId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointhostedzoneidParameter400E20C0,
        },
        Name: {
          Fn::Join: [
            ,
            [
              html-to-pdf.,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointssmdomainnameParameterD95B813B,
              },
              .,
            ],
          ],
        },
        Type: A,
      },
      Type: AWS::Route53::RecordSet,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointapistage8C7F5C96: {
      DependsOn: [
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiAccountED1DC265,
      ],
      Properties: {
        DeploymentId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointdeployment_UNIQUE_HASH,
        },
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
        StageName: default,
      },
      Type: AWS::ApiGateway::Stage,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4: {
      Properties: {
        DomainName: {
          Fn::Join: [
            ,
            [
              html-to-pdf.,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointssmdomainnameParameterD95B813B,
              },
            ],
          ],
        },
        EndpointConfiguration: {
          Types: [
            REGIONAL,
          ],
        },
        RegionalCertificateArn: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointorgformationcertswildcardcert1arnParameter1AF8E75E,
        },
        SecurityPolicy: TLS_1_2,
      },
      Type: AWS::ApiGateway::DomainName,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomainMapMyTestStackhtmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi730DB893D68EDB9C: {
      Properties: {
        DomainName: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4,
        },
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
        Stage: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentStageprod256692E7,
        },
      },
      Type: AWS::ApiGateway::BasePathMapping,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointdeployment_UNIQUE_HASH: {
      DependsOn: [
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyANY05444A22,
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiANY4A812312,
      ],
      Properties: {
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
      },
      Type: AWS::ApiGateway::Deployment,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointhtmltopdfsvcdomainnameEB41BF7D: {
      Properties: {
        Description: html-to-pdf service domain name,
        Name: HTML_TO_PDF_SVC_DOMAIN_NAME,
        Type: String,
        Value: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointcustomdomain808804C4,
        },
      },
      Type: AWS::SSM::Parameter,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203: {
      Properties: {
        Description: html-to-pdf Service API Gateway,
        Name: html-to-pdf RestAPI,
      },
      Type: AWS::ApiGateway::RestApi,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiANY4A812312: {
      Properties: {
        AuthorizationType: NONE,
        HttpMethod: ANY,
        Integration: {
          IntegrationHttpMethod: POST,
          Type: AWS_PROXY,
          Uri: {
            Fn::Join: [
              ,
              [
                arn:,
                {
                  Ref: AWS::Partition,
                },
                :apigateway:,
                {
                  Ref: AWS::Region,
                },
                :lambda:path/2015-03-31/functions/,
                {
                  Fn::GetAtt: [
                    htmltopdfstacknestjsNodeJsLambda82EBD38C,
                    Arn,
                  ],
                },
                /invocations,
              ],
            ],
          },
        },
        ResourceId: {
          Fn::GetAtt: [
            htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
            RootResourceId,
          ],
        },
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
      },
      Type: AWS::ApiGateway::Method,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiANYApiPermissionMyTestStackhtmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi730DB893ANYCDC44113: {
      Properties: {
        Action: lambda:InvokeFunction,
        FunctionName: {
          Fn::GetAtt: [
            htmltopdfstacknestjsNodeJsLambda82EBD38C,
            Arn,
          ],
        },
        Principal: apigateway.amazonaws.com,
        SourceArn: {
          Fn::Join: [
            ,
            [
              arn:,
              {
                Ref: AWS::Partition,
              },
              :execute-api:,
              {
                Ref: AWS::Region,
              },
              :,
              {
                Ref: AWS::AccountId,
              },
              :,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
              },
              /,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentStageprod256692E7,
              },
              /*/,
            ],
          ],
        },
      },
      Type: AWS::Lambda::Permission,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiANYApiPermissionTestMyTestStackhtmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi730DB893ANY1A7339A0: {
      Properties: {
        Action: lambda:InvokeFunction,
        FunctionName: {
          Fn::GetAtt: [
            htmltopdfstacknestjsNodeJsLambda82EBD38C,
            Arn,
          ],
        },
        Principal: apigateway.amazonaws.com,
        SourceArn: {
          Fn::Join: [
            ,
            [
              arn:,
              {
                Ref: AWS::Partition,
              },
              :execute-api:,
              {
                Ref: AWS::Region,
              },
              :,
              {
                Ref: AWS::AccountId,
              },
              :,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
              },
              /test-invoke-stage/*/,
            ],
          ],
        },
      },
      Type: AWS::Lambda::Permission,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiAccountED1DC265: {
      DeletionPolicy: Retain,
      DependsOn: [
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
      ],
      Properties: {
        CloudWatchRoleArn: {
          Fn::GetAtt: [
            htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiCloudWatchRole36037177,
            Arn,
          ],
        },
      },
      Type: AWS::ApiGateway::Account,
      UpdateReplacePolicy: Retain,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiCloudWatchRole36037177: {
      DeletionPolicy: Retain,
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Action: sts:AssumeRole,
              Effect: Allow,
              Principal: {
                Service: apigateway.amazonaws.com,
              },
            },
          ],
          Version: 2012-10-17,
        },
        ManagedPolicyArns: [
          {
            Fn::Join: [
              ,
              [
                arn:,
                {
                  Ref: AWS::Partition,
                },
                :iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs,
              ],
            ],
          },
        ],
      },
      Type: AWS::IAM::Role,
      UpdateReplacePolicy: Retain,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentDAC5D5D3bb7a2011d30d6d63955698ae6b66043e: {
      DependsOn: [
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyANY05444A22,
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyB06587AE,
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiANY4A812312,
      ],
      Properties: {
        Description: html-to-pdf Service API Gateway,
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
      },
      Type: AWS::ApiGateway::Deployment,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentStageprod256692E7: {
      DependsOn: [
        htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiAccountED1DC265,
      ],
      Properties: {
        DeploymentId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentDAC5D5D3bb7a2011d30d6d63955698ae6b66043e,
        },
        MethodSettings: [
          {
            DataTraceEnabled: false,
            HttpMethod: *,
            ResourcePath: /*,
            ThrottlingBurstLimit: 10,
            ThrottlingRateLimit: 10,
          },
        ],
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
        StageName: prod,
      },
      Type: AWS::ApiGateway::Stage,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyANY05444A22: {
      Properties: {
        AuthorizationType: NONE,
        HttpMethod: ANY,
        Integration: {
          IntegrationHttpMethod: POST,
          Type: AWS_PROXY,
          Uri: {
            Fn::Join: [
              ,
              [
                arn:,
                {
                  Ref: AWS::Partition,
                },
                :apigateway:,
                {
                  Ref: AWS::Region,
                },
                :lambda:path/2015-03-31/functions/,
                {
                  Fn::GetAtt: [
                    htmltopdfstacknestjsNodeJsLambda82EBD38C,
                    Arn,
                  ],
                },
                /invocations,
              ],
            ],
          },
        },
        ResourceId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyB06587AE,
        },
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
      },
      Type: AWS::ApiGateway::Method,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyANYApiPermissionMyTestStackhtmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi730DB893ANYproxyEEAA420F: {
      Properties: {
        Action: lambda:InvokeFunction,
        FunctionName: {
          Fn::GetAtt: [
            htmltopdfstacknestjsNodeJsLambda82EBD38C,
            Arn,
          ],
        },
        Principal: apigateway.amazonaws.com,
        SourceArn: {
          Fn::Join: [
            ,
            [
              arn:,
              {
                Ref: AWS::Partition,
              },
              :execute-api:,
              {
                Ref: AWS::Region,
              },
              :,
              {
                Ref: AWS::AccountId,
              },
              :,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
              },
              /,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiDeploymentStageprod256692E7,
              },
              /*/*,
            ],
          ],
        },
      },
      Type: AWS::Lambda::Permission,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyANYApiPermissionTestMyTestStackhtmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi730DB893ANYproxyEA6B4324: {
      Properties: {
        Action: lambda:InvokeFunction,
        FunctionName: {
          Fn::GetAtt: [
            htmltopdfstacknestjsNodeJsLambda82EBD38C,
            Arn,
          ],
        },
        Principal: apigateway.amazonaws.com,
        SourceArn: {
          Fn::Join: [
            ,
            [
              arn:,
              {
                Ref: AWS::Partition,
              },
              :execute-api:,
              {
                Ref: AWS::Region,
              },
              :,
              {
                Ref: AWS::AccountId,
              },
              :,
              {
                Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
              },
              /test-invoke-stage/*/*,
            ],
          ],
        },
      },
      Type: AWS::Lambda::Permission,
    },
    htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapiproxyB06587AE: {
      Properties: {
        ParentId: {
          Fn::GetAtt: [
            htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
            RootResourceId,
          ],
        },
        PathPart: {proxy+},
        RestApiId: {
          Ref: htmltopdfstackhtmltopdfstackapiendpointhtmltopdfstackapiendpointrestapi1B430203,
        },
      },
      Type: AWS::ApiGateway::Resource,
    },
    htmltopdfstacknestjsNodeJsLambda82EBD38C: {
      DependsOn: [
        htmltopdfstacknestjsNodeJsLambdaServiceRole59DBD5B4,
      ],
      Properties: {
        Code: {
          S3Bucket: {
            Fn::Sub: cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region},
          },
          S3Key: UNIQUE_HASH.zip,
        },
        Environment: {
          Variables: {
            AWS_ACCOUNT_ID: {
              Ref: AWS::AccountId,
            },
            MONGO_DATABASE_PASSWORD: {
              Ref: mongodatabasepasswordssmParameter,
            },
            MONGO_DATABASE_URI: {
              Ref: mongodatabaseurissmParameter,
            },
            MONGO_DATABASE_USER: {
              Ref: mongodatabaseuserssmParameter,
            },
            STAGE_NAME: default,
          },
        },
        Handler: index.handler,
        Layers: [
          {
            Ref: ChromiumLayer505381F6,
          },
          {
            Ref: lambdalayernestjslatestversionssmParameter,
          },
        ],
        MemorySize: 1600,
        Role: {
          Fn::GetAtt: [
            htmltopdfstacknestjsNodeJsLambdaServiceRole59DBD5B4,
            Arn,
          ],
        },
        Runtime: nodejs18.x,
        Timeout: 30,
      },
      Type: AWS::Lambda::Function,
    },
    htmltopdfstacknestjsNodeJsLambdaLogRetention0C09778D: {
      Properties: {
        LogGroupName: {
          Fn::Join: [
            ,
            [
              /aws/lambda/,
              {
                Ref: htmltopdfstacknestjsNodeJsLambda82EBD38C,
              },
            ],
          ],
        },
        RetentionInDays: 1,
        ServiceToken: {
          Fn::GetAtt: [
            LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A,
            Arn,
          ],
        },
      },
      Type: Custom::LogRetention,
    },
    htmltopdfstacknestjsNodeJsLambdaServiceRole59DBD5B4: {
      Properties: {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Action: sts:AssumeRole,
              Effect: Allow,
              Principal: {
                Service: lambda.amazonaws.com,
              },
            },
          ],
          Version: 2012-10-17,
        },
        ManagedPolicyArns: [
          {
            Fn::Join: [
              ,
              [
                arn:,
                {
                  Ref: AWS::Partition,
                },
                :iam::aws:policy/service-role/AWSLambdaBasicExecutionRole,
              ],
            ],
          },
        ],
      },
      Type: AWS::IAM::Role,
    },
  },
  Rules: {
    CheckBootstrapVersion: {
      Assertions: [
        {
          Assert: {
            Fn::Not: [
              {
                Fn::Contains: [
                  [
                    1,
                    2,
                    3,
                    4,
                    5,
                  ],
                  {
                    Ref: BootstrapVersion,
                  },
                ],
              },
            ],
          },
          AssertDescription: CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.,
        },
      ],
    },
  },
}
`;
